Add exporting mod lists to text files

Trial97 opened the pull request:
<!--
Hey there! Thanks for your contribution.

Please make sure that your commits are signed off first.
If you don't know how that works, check out our contribution guidelines: https://github.com/PrismLauncher/PrismLauncher/blob/develop/CONTRIBUTING.md#signing-your-work
If you already created your commits, you can run `git rebase --signoff develop` to retroactively sign-off all your commits and `git push --force` to override what you have pushed already.

Note that signing and signing-off are two different things!
-->
fixes #1152
This has just the mods to text (~~no curseforge export for the moment~~ that's implemented in another [PR](https://github.com/PrismLauncher/PrismLauncher/pull/1249) )

Trial97 committed:
Added ExportModsToStringTask

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
Added ExportModsToStringDialog

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
Fixed trigger function

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
Fixed code quality

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
Use slug for url

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
Updated slug for url

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

DioEgizio commented:
few issues:
- <img width="762" alt="image" src="https://github.com/PrismLauncher/PrismLauncher/assets/83089242/b92a2383-359b-4dcb-8d1c-569f1ec173e4"> this :P
- i think that it should be renamed from "text" to "Mod List" because all you're exporting is a mod list
- should probably support plaintext too
- it should have a save button where to save the html/md/plaintext
- should probably also render the md and the html somewhere

Trial97 committed:
Merge branch 'develop' of https://github.com/PrismLauncher/PrismLauncher into export

Trial97 committed:
Added plantxt export

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
Added markdown QT version check

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
Use internal markdown implementation

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

DioEgizio commented on launcher/ui/MainWindow.ui:
```diff
@@ -479,6 +479,14 @@
+    <iconset theme="new"/>
+   </property>
+   <property name="text">
+    <string>ModList (txt)</string>
```
```suggestion
    <string>Mod List</string>
```

DioEgizio commented on launcher/ui/dialogs/ExportToModListDialog.cpp:
```diff
@@ -0,0 +1,172 @@
+    QString exampleLine;
+    switch (format) {
+        case ExportToModList::HTML: {
+            exampleLine = "<ul><a href=\"{url}\">{name}</a>[{version}] by {authors}</ul>";
```
```suggestion
            exampleLine = "<ul><a href=\"{url}\">{name}</a> [{version}] by {authors}</ul>";
```

DioEgizio commented on launcher/ui/dialogs/ExportToModListDialog.cpp:
```diff
@@ -0,0 +1,172 @@
+            break;
+        }
+        case ExportToModList::MARKDOWN: {
+            exampleLine = "[{name}]({url})[{version}] by {authors}";
```
```suggestion
            exampleLine = "[{name}]({url}) [{version}] by {authors}";
```

DioEgizio commented on launcher/ui/dialogs/ExportToModListDialog.cpp:
```diff
@@ -0,0 +1,172 @@
+            break;
+        }
+        case ExportToModList::PLAINTXT: {
+            exampleLine = "name: {name}; url: {url}; version: {version}; authors: {authors};";
```
```suggestion
            exampleLine = "{name} ({url}) [{version}] by {authors}";
```

DioEgizio reviewed (commented):
<img width="1222" alt="image" src="https://github.com/PrismLauncher/PrismLauncher/assets/83089242/3235b50b-0554-4f1a-af78-2fe303624f2f">
also the html renderer doesn't seem to correctly linebreak

Trial97 committed:
Update launcher/ui/MainWindow.ui

Co-authored-by: DioEgizio <83089242+DioEgizio@users.noreply.github.com>
Signed-off-by: Alexandru Ionut Tripon <alexandru.tripon97@gmail.com>

Trial97 committed:
Update launcher/ui/dialogs/ExportToModListDialog.cpp

Co-authored-by: DioEgizio <83089242+DioEgizio@users.noreply.github.com>
Signed-off-by: Alexandru Ionut Tripon <alexandru.tripon97@gmail.com>

Trial97 committed:
Update launcher/ui/dialogs/ExportToModListDialog.cpp

Co-authored-by: DioEgizio <83089242+DioEgizio@users.noreply.github.com>
Signed-off-by: Alexandru Ionut Tripon <alexandru.tripon97@gmail.com>

Trial97 committed:
Update launcher/ui/dialogs/ExportToModListDialog.cpp

Co-authored-by: DioEgizio <83089242+DioEgizio@users.noreply.github.com>
Signed-off-by: Alexandru Ionut Tripon <alexandru.tripon97@gmail.com>

Scrumplex commented:
This is a weird feature. I don't really see the use case if people can just create mrpacks now anyway.

DioEgizio commented:
I can see the use of it tbh. You don't want an entire modpack for just listing your mods. It can be useful also for example to modpack creators that want to list their mod in their description. With this they can just export- mod list- markdown and they can just copy paste it (same for html on CF). I can see many uses. Not to mention that some mods are on cf and some on modrinth

Trial97 committed:
fix: html sintax for modlist export

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
If not desired feel free to close this but also close the linked issue with an explanation of why this is not wanted

DioEgizio commented:
I think this is useful in many scenarios tbh

Scrumplex commented:
Well. This is something that can be automated using shell scripts. Like, I don't see how this feature fits in the context of a launcher.

DioEgizio commented:
Shell scripts parsing metadata information?

Scrumplex commented:
> Shell scripts parsing metadata information?

You can use tools like `yq` to read the Packwiz TOML files.

> It can be useful also for example to modpack creators that want to list their mod in their description

I haven't seen many modpacks do that (usually because the mod platform will already do it for you mostly)

---

I just don't think this feature belongs into the launcher. I really hate this word, but I think "bloat" fits well here. If we add niche features like this, we only increase the amount of maintenance needed in the future.

Also, where does this feature end? Are we gonna need an HTML and a BBCode export as well? Maybe someone needs to post the modlist in a Minecraft Forums post, which doesn't support Markdown

Trial97 commented:
> Also, where does this feature end? Are we gonna need an HTML and a BBCode export as well? Maybe someone needs to post the modlist in a Minecraft Forums post, which doesn't support Markdown

Well, this already supports HTML and for other stuff, I added a custom format that lets the user define how each line should look.

DioEgizio commented:
> You can use tools like yq to read the Packwiz TOML files.

It wouldn't need just to to that, but to read the metadata inside the mod file. At that point it's not even a shell scripts, it's a mod manager

> I haven't seen many modpacks do that (usually because the mod platform will already do it for you mostly)

Some do

Scrumplex commented:
Look. I usually try to look at the problem in an issue, instead of accepting their proposed solution to fix their problem. The issue states the following benefits:

> - Easy Sharing: Users can share the exported mod lists with others, making it simpler to discuss and compare mods.
> - Documentation: Exported mod lists can serve as documentation for reference or troubleshooting purposes.

Both of these can be achieved using the mrpack export. So why add another feature for the same purpose?

DioEgizio commented:
mrpack export can only be used for, well, modrinth mods and mods that can be redistributed. It doesn't work for CF only mods and feels redundant for just a mod list. Like imagine if someone asks you what mods you use with the links to them and you literally have to make an entire modpack. Not to mention you also might want to add the mods to an already existing instance

Scrumplex commented:
My problem with this is that this is intended to output to a human-readable format. You can't use an exported mod list to import it into another instance anyway. It would be a lot of manual work.

If you want to share your mods with a friend you could send them the list of the jar files, you could take a screenshot of your mods either in Prism or even in-game.

DioEgizio commented:
Still, I don't understand what's the problem with a human readable mod list. It's the reason why things like moddermore exist after all

Scrumplex commented:
The reasons against are 700 additional lines of code that have to be maintained.

DioEgizio commented:
I mean it's not like this is the only feature that could be considered niche (but I don't think it is tbh). We literally have a java profiler option

Trial97 committed:
Added json and csv format

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
Now this would also implement the export options mentioned in the following issues https://github.com/PrismLauncher/PrismLauncher/issues/214, https://github.com/PrismLauncher/PrismLauncher/issues/1178 (no importing for this)
I'm linking them here so that they can be addressed once this PR is closed.

Trial97 committed:
Moved to separate functions

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
escaped text for html export

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

TheKodeToad commented:
This is quite nice!

Maybe it should prompt to update/"sync" the mods?
I found a lot of URLs were missing...

TheKodeToad commented:
Ohh - also maybe `Optional Info` could change into buttons to insert the placeholders on the Custom option.

And instead of preventing the Template from being edited, you could detect when it's edited and switch to custom...

Trial97 committed:
Made custom template enabled all time

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
Thanks for the feedback:
- instead of prompting to update mods, I added a message that should explain that the metadata is needed
- updated the template behavior 
- regarding Oprional info I can't really picture what you mean so I kept it as is

TheKodeToad commented:
Maybe have buttons to insert {author} etc...

Trial97 commented:
I don't know what to say about this...
If I go this route, I should remove the extra file types as they do not make sense anymore if the buttons change the template.
In that case, I would only have a custom and a template, is that what you want?

TheKodeToad commented:
Oops... no, I meant only have that behaviour in custom

Trial97 committed:
Added buttons

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
This should solve it then

TheKodeToad commented:
Nice :3

I'll try to properly review the changes now if i don't get distracted yet again :P

TheKodeToad commented on launcher/modplatform/helpers/ExportToModList.cpp:
```diff
@@ -0,0 +1,200 @@
+    return lines.join("\n");
+}
+
+QString ExportToModList(QList<Mod*> mods, Formats format, OptionalData extraData)
```
```suggestion
QString exportToModList(QList<Mod*> mods, Formats format, OptionalData extraData)
```

TheKodeToad commented on launcher/modplatform/helpers/ExportToModList.cpp:
```diff
@@ -0,0 +1,200 @@
+    }
+}
+
+QString ExportToModList(QList<Mod*> mods, QString lineTemplate)
```
```suggestion
QString exportToModList(QList<Mod*> mods, QString lineTemplate)
```

TheKodeToad commented on launcher/modplatform/helpers/ExportToModList.cpp:
```diff
@@ -0,0 +1,200 @@
+    return QString("<html><body><ul>\n\t%1\n</ul></body></html>").arg(lines.join("\n\t"));
+}
+
+QString toMARKDOWN(QList<Mod*> mods, OptionalData extraData)
```
ARGHHHHHHHHHH!!!!!!!!!!

I think Markdown would make more sense :laughing:

TheKodeToad commented on launcher/modplatform/helpers/ExportToModList.cpp:
```diff
@@ -0,0 +1,200 @@
+    return lines.join("\n");
+}
+
+QString toPLAINTXT(QList<Mod*> mods, OptionalData extraData)
```
PlainTXT?

TheKodeToad reviewed (changes requested):
I'm confused what the changes to dependencies are for...

Trial97 commented:
> I'm confused what the changes to dependencies are for...

oops I think I merged the wrong develop(I will remove that)

Trial97 committed:
renames

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
Merge branch 'develop' of https://github.com/PrismLauncher/PrismLauncher into export

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
Ok now it should be merged ok

TheKodeToad commented on launcher/ui/dialogs/ExportToModListDialog.ui:
```diff
@@ -0,0 +1,240 @@
+     <item>
+      <widget class="QLabel" name="warningLabel">
+       <property name="text">
+        <string>This depends on the mods meta data. To ensure the meta data run at least one time the mods update on the selected instance(no need to update the mods).</string>
```
```suggestion
        <string>This depends on the mods' metadata. To ensure it is available, run an update on the instance. Installing the updates isn't necessary.</string>
```

Trial97 committed:
Update launcher/ui/dialogs/ExportToModListDialog.ui

Co-authored-by: TheKodeToad <TheKodeToad@proton.me>
Signed-off-by: Alexandru Ionut Tripon <alexandru.tripon97@gmail.com>

TheKodeToad reviewed (approved):
good! ty

Scrumplex commented:
I still think this is an overkill feature, but eh

ruguysgoingtrickortreating commented:
Hi, I stumbled across this pull request because my enhancement was mentioned, so would it be possible to import the file this generates and automatically download the mods from their respective repositories like the download mods menu? It would add a lot of use to this feature.

Scrumplex commented:
> Hi, I stumbled across this pull request because my enhancement was mentioned, so would it be possible to import the file this generates and automatically download the mods from their respective repositories like the download mods menu? It would add a lot of use to this feature.

You can already do this using the Modrinth export feature. Just export your instance as a Modrinth modpack and import it somewhere else.

DioEgizio commented:
Same with the cf export but I think he's talking about when mods have different types providers
